File: 1_OneWeekend/1_Graphics_Hello_World/cpp/ppm.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

bool outputFile(std::string const &fileName, 
                std::string const &header,
                std::vector<std::string> const &pixels) {
    std::ofstream output;
    try {
        output.open(fileName);
        output << header.c_str();
        for (auto rgb : pixels) {
            output << rgb.c_str();
        }
        return true;
    } catch (...) {
        std::cout << "Any error occured writing image" << std::endl;
    }
    return false;
}

int main() {
    int nx = 200;
    int ny = 100;
    std::vector<std::string> pixels;
    pixels.reserve(nx*ny);
    std::string header = "P3\n" + std::to_string(nx) + " " + std::to_string(ny) + "\n255\n";
    std::string rgb;
    for (int j = ny-1; j >= 0; j--) {
        for (int i = 0; i < nx; i++) {
            float r = float(i) / float(nx);
            float g = float(j) / float(ny);
            float b = 0.2;
            int ir = int(255.99*r);
            int ig = int(255.99*g);
            int ib = int(255.99*b);
            rgb = std::to_string(ir) + " " + std::to_string(ig) + " " + std::to_string(ib) + "\n";
            pixels.emplace_back(rgb);
        }
    }
    outputFile("cppimage.ppm", header, pixels);
}

File: 1_OneWeekend/1_Graphics_Hello_World/nim/ppm.nim

proc outputFile(fileName: string, header: string, pixels: seq[string]): bool =
  var output = open(fileName, fmWrite)
  try:
    output.writeLine(header)
    for p in pixels:
      output.writeLine(p)
    return true
  except:
    echo "Any error occured writing image"
  return false

proc generateHeader(x: int, y: int): string =
  var header: string = "P3\n" & $x & " " & $y & "\n255"
  return header

proc generatePixels(x: int, y: int): seq[string] =
  let nx = x
  let ny = y
  var pixels = newSeq[string]()
  var rgb: string

  for j in countdown(ny-1, 0):
    for i in 0..<nx:
      let r: float = toFloat(i) / toFloat(nx)
      let g: float = toFloat(j) / toFloat(ny)
      let b: float = 0.2
      let ir: int = toInt(255.99*r)
      let ig: int = toInt(255.99*g)
      let ib: int = toInt(255.99*b)
      rgb = $ir & " " & $ig & " " & $ib
      pixels.add(rgb)

  return pixels

# main
let x = 200
let y = 100
let header = generateHeader(x, y)
let pixels = generatePixels(x, y)
let image = outputFile("nimimage.ppm", header, pixels)
if image == false:
  echo "Error writing image file"
